### ISSUE TITLE
Proposal: Variable Memory Limits for Queries

### ISSUE THREAD

**Issue by jsternberg** [influxdata/influxdb#15265]
Variable Memory Limits
----------------------

### Problem

The Flux engine provides support for tracking the memory usage of each query and provides the ability to restrict the amount of memory that a query may use before it gets killed.

This principle is a cornerstone for how we plan to allow for multi-tenant functionality and long term scalability of the system.

On the other hand, at the moment, we are not reaching that goal. The reason for this is because our memory limits are very simplistic and inelastic. They are unable to adapt to a living system because we only have hard limits and those hard limits dictate a lot about how we set the memory limits.

When calculating how many queries are able to run concurrently, we have to consider a few aspects.

1.	How much memory is each query allowed to use?
2.	How much overhead of Flux runtime memory will be used by the Go runtime and is thus untracked?
3.	How many queries are allowed to execute at the same time?

The first and third are the most important. Because our memory limits are set at the beginning of the runtime and never change, we have to be certain that the amount of memory that the query will use is actually available for use at the time the query executes. This ends up limiting the amount of queries that can be run concurrently because we need to ensure that `N * C` amount of memory is used with `N` being the amount of memory per query and `C` being the concurrency. The value of `N` ends up getting set to a value that we are certain close to 100% of queries will be able to pass through regardless of how much memory may actually get used by the query. This greatly limits the concurrency of the system because we run into one of two situations:

1.	We are unable to feel safe that the memory limits will prevent OOMs because the concurrency multiplied by the memory per query exceeds the memory of the container running the query.
2.	We are unable to set the concurrency to a higher value and we both waste CPU that we are paying for and waste memory that we are paying for because most queries will never use the amount of memory reserved for them.

For this reason, we need a way to allow for memory limits to be changed during the execution of a query. The intention behind this is to be able to reserve a certain amount of memory for each query. This number will be a value that the 99th percentile of queries will require. That means that 99% of queries will only ever use their initial allocation of memory and will never have to request more. The other 1% of queries will still be able to succeed, but may fail if the system is memory starved.

### Proposal

The proposal is to modify the `memory.Allocator` to take an optional `memory.Manager` interface. This interface will be used by the `memory.Allocator` to request more memory so it can raise its own `Limit`. The `Limit` that is set in the `memory.Allocator` will be the currently reserved memory.

```go
package memory

type Manager interface {
    // RequestMemory will request that the given amount of memory
    // be reserved for the caller. The manager will return the number
    // of bytes that were successfully reserved. The n value will be
    // either equal to or greater than the requested number of bytes.
    // If the manager cannot reserve at least bytes in memory, then
    // it will return an error with the details.
    RequestMemory(bytes int64) (n int64, err error)

    // FreeMemory will inform the memory manager that this memory
    // is not being used anymore.
    // It is not required for this to be called similar to how
    // it is not necessary for a program to free the memory.
    // It is the responsibility of the controller to reclaim the
    // unfreed memory when a resource is dead.
    FreeMemory(bytes int64)
}

type Allocator struct {
    // Limit is the limit on the amount of memory that this allocator
    // can assign. If this is null, there is no limit.
    Limit *int64

    // Manager holds the Manager for this Allocator.
    // If this Allocator has a limit set and the limit is to be exceeded,
    // it will attempt to use the Manager to request more memory.
    // If this fails, then the Allocator will panic.
    Manager Manager

    // Allocator is the underlying memory allocator used to
    // allocate and free memory.
    // If this is unset, the DefaultAllocator is used.
    Allocator memory.Allocator
}
```

The advantages of this method are both its simplicity and its flexibility. This method would allow a query to grow the amount of memory it is allowed to use while also allowing a decision to be made to forbid a greater amount of memory to be used. Additionally, because of the structure, a query would be forced to inform the controller about the change in memory expectations which would give the controller the ability to forcibly kill the query under undesirable circumstances. It would also be possible to implement other types of limits with this model.

### Future Work

In the future, the `memory.Allocator` struct will be an implementation detail of the controller in InfluxDB or whatever other system embeds Flux. The Flux engine will not concern itself with the inner implementations of how the memory allocator works and it will use the common interface already present in the Apache Arrow library.

We are unable to do this at the moment though because certain code inside of Flux does not use Arrow for memory allocations. It manually tracks its own memory usage from slices using the custom `execute.Allocator` interface. Once that is removed, we will no longer need `memory.Allocator` to be exposed and it can be a formal part of the controller.

**Comment by nathanielc** [comment]
This looks good. The important parts to me are this allows for more efficient use of resources while still being very simple. 

In reading through this is seems like we are making  slight trade off between noisy neighbors and efficiency. In being more efficient we are closer to the upper bound of the raw resources. This means that a noisy neighbor has more of an opportunity to cross that threshold since its a smaller gap. It seems like we have a knob we can tune between efficiency and noisy neighbors. 

This change allows us to experiment with that knob and learn more about the behavior of our system. Its also a simple change so if we discover it has negative results its easy to revert.

**Comment by jsternberg** [comment]
This is implemented and will be merged when we release v0.50.0 of flux to influxdb. It relies on that version so it is currently merged to that staging branch. I'm going to close this issue since we no longer need it for tracking the work.
